app/src/features/surveys/components/StudyAreaForm.tsx

```diff
@@ -46,25 +49,88 @@ export const SurveyLocationYupSchema = yup.object({
  */
 const StudyAreaForm = () => {
   const formikProps = useFormikContext<ISurveyLocationForm>();
+  const { handleSubmit, values, setFieldValue } = formikProps;
+  const [isOpen, setIsOpen] = useState(false);
+  const [currentIndex, setCurrentIndex] = useState<number | undefined>(undefined);
+  const drawRef = createRef<IDrawControlsRef>();
+
+  const onOpen = () => {
+    setIsOpen(true);
+  };
+  const onClose = () => {
+    setIsOpen(false);
+    setCurrentIndex(undefined);
+  };
+  const onSave = (data: { name: string; description: string }) => {
+    setFieldValue(`locations[${currentIndex}].name`, data.name);
+    setFieldValue(`locations[${currentIndex}].description`, data.description);
+  };
+
+  const onDelete = (index: number) => {
+    // remove the item at index
+    const data = values.locations;
+    const locationData = data.splice(index, 1);
+
+    // Use Draw Ref to remove editable layer form the map
+    locationData.forEach((item) => {
+      if (item.leaflet_id) {
+        drawRef.current?.deleteLayer(item.leaflet_id);
+      }
+    });
+
+    // set values
+    setFieldValue('locations', data);
+  };
+
+  const getDialogData = () => {
```

Comment:
This should probably be a constant, rather than a callback
```javascript
const dialogInitialValues = useMemo(() => {
  if (currentIndex === undefined) {
    return { name: '', description: '' };
  }

  return {
    name: values.locations[currentIndex]?.name,
    description: values.locations[currentIndex]?.description
  };
}, [values.locations, currentIndex]);
```

Reply:
the useMemo here is a callback? The value also isn't so large that caching will provide much value, I put this into a function to avoid having the logic inline for the dialog.

Reply:
(The `useMemo` isn't a callback, but it uses a callback to produce a constant)

Agreeing to your point of not inlining this logic in the React component, what I really meant is that having a function `getDialogData()` that always returns the same object should really just be replaced with an object. I think it really makes more sense to just instantiate the constant. To that end, it's really best practice to memoize _any_ variable that is instantiated during component render, hence the `useMemo`.

Reply:
Ooh I see. Ok I've updated this chunk to be inline with best practices